arch/x86/adler32_avx512_tpl.h

```diff
@@ -46,10 +46,10 @@ Z_INTERNAL uint32_t adler32_avx512(uint32_t adler, const uint8_t *src, uint64_t
 
     __m512i vbuf, vs1_0, vs3;
 
-    const __m512i dot2v = _mm512_set_epi8(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
-                                          20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37,
-                                          38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55,
-                                          56, 57, 58, 59, 60, 61, 62, 63, 64);
+    const __m512i dot2v = _mm512_set_epi32(0x1020304, 0x5060708, 0x90a0b0c, 0xd0e0f10,
```

Comment:
While I'm sure this code is functionally correct - I do not love that it impedes legibility. Is there any possible way to write a macro that uses a fallback method to derive these same constants at compile time? That just makes it much more obvious what's going on and if someone wants to go back and change the implementation to say, use a different mixing matrix of sorts, they don't have to struggle using the set_epi32 intrinsic to derive the proper constants.

Reply:
There's also the fact that GCC prior to 9 kind of emitted garbage code for some AVX512 intrinsics, but that's another matter.  It was missing a lot of keyhole optimizations with regard to masks (and it's still not great, it zealously moves masks into GPRs a bit too often).

Reply:
What do you think about doing #ifdef __GNUC__ < 9 with both versions in the source?

Reply:
Doable, but it might be cleaner and more future proof to just have selectively bit shifts per byte (and 16 bitwise ORs) on gcc versions where the intrinsic isn't available.  Getting that right means we should be able to specify exact uint8 values per lane should the implementation change or AVX512 gets used elsewhere.

Reply:
We have header(s) for "fallback" functions for a reason... If compiler doesn't support something, we should add any replacement function to most relevant fallback header. 

All fallback functions should be generic with no expectations on what parameters are passed to them. Any repeated calculations should be added as macros to make the code as readable as possible. As there is only little-endian x86 processors, there only needs to be one variation of the macros. Care has to be taken with preprocessor macros as normal integer promotion and conversion rules do not apply, so shifting must mask any extra bits.